### DISCUSSION TITLE
Python script optional parameters and ui generation

### DISCUSSION THREAD

**Discussion by ilazarte** [windmill-labs/windmill#1784] [Questions]
How do I define an optional parameter for Python scripts?  The form I prefer: 
```
def foo(first_name: str = None):
    return f'Hello world from {first_name}'
```

doesn't appear to be recognized by the script parser correctly and shows red squiggles.  I'm running the latest I assume (self host setup script)

edit: for reference I also tried Optional[str] from the typing module.  It does make it optional but then requires quotes around the string in the input box which I find weird, it should understand it as either completely none, or a valid str.  I did notice the ui thinks it's an `object` after using the Optional form.

**Comment by ghost** [comment]
This seems to work now (although I don't know about the web editor).

My version at this time: v1.349.1

Using `wmill script generate-metadata` brings up a field that is not required as expected.

Using `typing.Optional` completely fails here:
```
$ wmill script generate-metadata f/transform/aggregate-data.py 
Generating metadata for f/transform/aggregate-data.py
Script f/transform/aggregate-data.py invalid, it cannot be parsed to infer schema.
^C
```
^^ note that it just hangs there and I needed to interrupt it.

**Reply by rubenfiszel** [reply]
@enote-kane could you make an issue for both the hanging and for us supporting Optional, we should be able to do those

**Reply by ghost** [reply]
@rubenfiszel Turned out that it was my fault, since I had a parameter with a default value before the optional, so like this:
```python
def main(required: str, with_def: str = "test", opt: Optional[str]):
    print("All OK")
```

That will not work (of course), since Python itself will complain about that definition (parameter `opt` has no default, hence will refuse to parse):
```
>>> def main(required: str, with_def: str = "test", opt: Optional[str]):
  File "<stdin>", line 1
    def main(required: str, with_def: str = "test", opt: Optional[str]):
                                                    ^^^^^^^^^^^^^^^^^^
SyntaxError: parameter without a default follows parameter with a default
```

Created #3954 for supporting `Optional`.
